src/core/Akka/Util/FastLazy.cs

```diff
@@ -0,0 +1,63 @@
+﻿using System;
+using System.Runtime.CompilerServices;
+using System.Threading;
+
+namespace Akka.Util
+{
+    /// <summary>
+    /// A fast, atomic lazy that only allows a single publish operation to happen,
+    /// but allows executions to occur concurrently.
+    /// 
+    /// Does not cache exceptions. Designed for use with <typeparam name="T"/> types that are <see cref="IDisposable"/>
+    /// or are otherwise considered to be expensive to allocate. 
+    /// 
+    /// Read the full explanation here: https://github.com/Aaronontheweb/FastAtomicLazy#rationale
+    /// </summary>
+    public sealed class FastLazy<T>
```

Comment:
Built tests, benchmarks, and docs for this in a separate repo: https://github.com/Aaronontheweb/FastAtomicLazy#rationale (mostly so I didn't clutter Akka.NET with more tests and benchmarks it doesn't need)